CPT: add TermTreeSelector to Editor Taxonomies

timmyc opened the pull request:
This branch brings the `<TermTreeSelector />` to the editor sidebar for custom post types that have a hierarchical taxonomy ( i.e. Portfolios ).

<img width="257" alt="edit_project_ _testingtimmy2_wordpress_com_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/22080/16543207/b6064d54-407e-11e6-8d2a-166a8b270253.png">

After working through the items in the _Prior Implementation Notes_ below, I opted to take a similar route we have done with categories in the post editor and to leverage the `normalizer()` in the `post-edit-store` here.  Still not sure if this is the optimal route, but having the data flow from the store->components in this manner feels less hacky then tracking selections in the component state.

__To Test__
- Open the post editor for a site that has Portfolios enabled
- Edit a portfolio or create a new portfolio via the sidebar link
- Interact with both Portfolio Types term selector and Portfolio Tags token field.  Update the post and verify that changes are persisted as expected

__Known Issue__
There is a known issue with the API where if you select two terms with the same name, the API since it returns terms keyed by name, will only return one.  This bug exists for categories currently as well.  So selecting two terms with the same name will not work properly.

__Prior Implementation Notes__
Still some kinks to work out here with this approach.  I encountered some issues with the API wrt using the `terms: {}` object to update hierarchical taxonomy selections.  Updating the terms object like it is received from the API and updating a post does not save the terms, and it produces a warning on my sandbox:

```js
{
terms: {
    jetpack_portfolio_type: {
        'Some Category Name': { ... }
    }
}
```

The `TokenField` implementation sets the term object like:

```
{
    terms: {
        jetpack_portfolio_tag: [ 'some term name', 'some other term name' ]
}
```

This is marshaled into an array on the API, which subsequently looks up the term by name and all works well.  I did attempt this approach with hierarchical taxonomies, and it does work( ish ), it will not work in scenarios where there are similar term names, at different levels of a tree:

![edit_project_ _testingtimmy2_wordpress_com_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/16510657/f7218944-3efc-11e6-9e4f-4575d8e8a0fe.png)

As such I went with `terms_by_id` here.  It is working well - though it doesn't seem as seamless as just reading/updating the `terms` object on the post.  There are some lingering issues too in this PR such as manipulating tags and types on a CPT during the same save request ( though should be easy enough to fix ).

For similar reasons I think this is what lead us to do hackery in the post-edit-store for categories.  It would be nice to find a solution so we don't need to revisit that approach again.

/cc @nylen @aduth 

Test live: https://calypso.live/?branch=add/cpt/term-tree-selector

aduth commented on client/post-editor/editor-drawer/taxonomies.jsx:
```diff
@@ -34,26 +35,29 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
 				}
+
+				return (
+					<Accordion
```
Since the render behavior is mostly consistent between the taxonomy types, aside from the content of the accordion and the Gridicon `icon`, maybe we could consolidate this a bit more?

```diff
iff --git a/client/post-editor/editor-drawer/taxonomies.jsx b/client/post-editor/editor-drawer/taxonomies.jsx
index 70a219e..4973ff7 100644
--- a/client/post-editor/editor-drawer/taxonomies.jsx
+++ b/client/post-editor/editor-drawer/taxonomies.jsx
@@ -25,37 +25,25 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
                                <QueryTaxonomies { ...{ siteId, postType } } />
                        ) }
                        { map( taxonomies, ( taxonomy ) => {
-                               if ( 'post_format' === taxonomy.name ) {
+                               const { name, label, hierarchical } = taxonomy;
+                               if ( 'post_format' === name ) {
                                        // Post format has its own dedicated accordion
                                        return;
                                }
 
-                               if ( taxonomy.hierarchical ) {
-                                       return (
-                                               <Accordion
-                                                       key={ taxonomy.name }
-                                                       title={ taxonomy.label }
-                                                       icon={ <Gridicon icon="folder" /> }
-                                               >
-                                                       <TermSelector
-                                                               selectedTermIds={ postTerms ? map( postTerms[ taxonomy.name ], 'ID' ) : [] }
-                                                               taxonomyName={ taxonomy.name }
-                                                       />
-                                               </Accordion>
-                                       );
-                               }
-
                                return (
                                        <Accordion
-                                               key={ taxonomy.name }
-                                               title={ taxonomy.label }
-                                               icon={ <Gridicon icon="tag" /> }
-                                       >
-                                               <TermTokenField
-                                                       postTerms={ postTerms }
-                                                       taxonomyName={ taxonomy.name }
-                                                       taxonomyLabel={ taxonomy.label }
-                                               />
+                                               key={ name }
+                                               title={ label }
+                                               icon={ <Gridicon icon={ hierarchical ? 'folder' : 'tag' } /> }>
+                                               { hierarchical
+                                                       ? <TermSelector
+                                                               selectedTermIds={ postTerms ? map( postTerms[ taxonomy.name ], 'ID' ) : [] }
+                                                               taxonomyName={ taxonomy.name } />
+                                                       : <TermTokenField
+                                                               postTerms={ postTerms }
+                                                               taxonomyName={ taxonomy.name }
+                                                               taxonomyLabel={ taxonomy.label } /> }
                                        </Accordion>
                                );
                        } ).filter( Boolean ) }
```

aduth commented on client/post-editor/editor-drawer/taxonomies.jsx:
```diff
@@ -34,26 +35,29 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
+						icon={ <Gridicon icon="tag" /> }
+					>
+						<TermTokenField
+							postTerms={ postTerms }
```
@nylen : Aside, maybe `postTerms` isn't the best name for the prop, since it seems to unnecessarily tie the field to being used only the context of editing a post. What about something with `selected` instead?

nylen replied:
These components are already pretty tightly tied to editing posts - [recording stats, `PostActions.edit`](https://github.com/Automattic/wp-calypso/blob/d98ba96676263eb9f1b6bc8d42c912af499a6d02/client/post-editor/term-token-field/index.jsx#L28-L47), and [pulling terms from a post terms object](https://github.com/Automattic/wp-calypso/blob/d98ba96676263eb9f1b6bc8d42c912af499a6d02/client/post-editor/term-token-field/index.jsx#L49-L61).  Take that away and you're left with just `TokenField`, so a component for another purpose should be written from scratch.

aduth commented on client/post-editor/editor-drawer/term-selector.jsx:
```diff
@@ -0,0 +1,80 @@
+/**
```
Should we move this to its own directory, `editor-term-selector` ? My thinking with files within `editor-drawer` is that they should be limited to the logic necessary to render the accordion itself, but using components from their own directories.

aduth commented on client/post-editor/editor-drawer/term-selector.jsx:
```diff
@@ -0,0 +1,80 @@
+
+		// TODO: REDUX - remove flux actions when whole post-editor is reduxified
+		PostActions.edit( {
+			terms_by_id: {
```
If I'm editing an existing post that has one term assigned and I add another, are both shown as selected? I'm not sure I follow that happening from the `postTerms` passed through from `<EditorDrawer />` (via `post.terms`, but `terms_by_id` editing here; guessing we'd want some normalization?).

aduth commented on client/post-editor/editor-drawer/term-selector.jsx:
```diff
@@ -0,0 +1,80 @@
+
+	onTermsChange( selectedTerm ) {
+		const { taxonomyName } = this.props;
+		let termIds = clone( this.state.selectedIds ) || [];
```
When will `selectedIds` be falsey to lead to the fallback? Looks from `<EditorDrawerTaxonomies />` that this won't happen. Otherwise, could we specify a default prop? 

In general, you could drop the clone in favor of:

```js
let termIds = [ ...this.state.selectedIds ];
```

But here, even better, you can avoid an unnecessary clone in the case that the term is already selected (`without` doesn't mutate) by using `Array#concat`:

```js
let termIds = this.state.selectedIds;
if ( -1 === termIds.indexOf( selectedTerm.ID ) ) {
	termIds = termIds.concat( selectedTerm.ID );
} else {
	termIds = without( termIds, selectedTerm.ID );
}

aduth commented on client/post-editor/editor-drawer/term-selector.jsx:
```diff
@@ -0,0 +1,80 @@
+			}
+		} );
+
+		this.setState( { selectedIds: termIds } );
```
If you renamed the `let` to `selectedIds` (which I don't think would be an inaccurate name), this could be changed to object shorthand:

```js
this.setState( { selectedIds } );
```

timmyc commented:
The prior approach had some flaws, as @aduth pointed out, but I wanted to get it up there as-is to show some of the challenges with manipulating a shared `terms` object.

In 98abac5 I have taken the route we already have using `normalize` in the `post-edit-store` to populate `term_ids`.  This removes the cruft of managing selected term IDs in the state.  Updates are now achieved similar to `term-token-field` by updating the `terms` object directly.

We still have the issue with hierarchical terms though where two terms at different points in the tree will conflict as we use the term name as a key in the object.  Feeling like we might have a bit of a bug there from the API too.

timmyc commented:
@aduth moved the component to its own directory and cleaned up logic in `taxonomies.jsx` in ec6885e.

kellychoffman commented:
🙏

aduth commented on client/post-editor/editor-drawer/taxonomies.jsx:
```diff
@@ -24,36 +25,26 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
+					>
+					{ hierarchical
+						? <TermSelector postTerms={ postTerms } taxonomyName={ name } />
+						: <TermTokenField postTerms={ postTerms } taxonomyName={ name } taxonomyLabel={ label } />
```
Another aside @nylen , we should use `getTerm` from within `<TermTokenField />`'s `connect` selector to retrieve the label, not force it to be passed in as a prop.

nylen replied:
→ #6547

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -0,0 +1,72 @@
+		const terms = cloneDeep( postTerms ) || {};
+		const taxonomyTerms = terms[ taxonomyName ];
+
+		if ( taxonomyTerms[ selectedTerm.name ] ) {
```
For a new post, `taxonomyTerms` will be `undefined` here, so trying to add or remove a term will result in an error.

>Uncaught TypeError: Cannot read property 'From API' of undefined

aduth replied:
I think it'd be sufficient to add `|| {}` fallback for `const taxonomyTerms`, and I was going to commit this, but encountered another issue where trying to remove a term from a saved CPT would not persist the removal on save.

timmyc commented:
> another issue where trying to remove a term from a saved CPT would not persist the removal on save.

I fear this might be an API bug.  The same issue happens when trying to remove all term tags from a saved post too.

timmyc commented:
Okay I think the issue with not being able to unset all previously saved terms might be due to [qs in wpcom.js](https://github.com/Automattic/wpcom.js/blob/master/lib/util/send-request.js#L68).  While debugging I would see that `terms_by_id` from the `post-edit-store` was correctly showing an empty array, but the subsequent request `PUT` via wpcom.js did not.

The following in a node console proved the suspicsion:

```
> qs.stringify({ 'jetpack-porfolio-type': [], 'jetpack-portfolio-tags': [ 1, 2 ] });
> 'jetpack-portfolio-tags%5B0%5D=1&jetpack-portfolio-tags%5B1%5D=2'
```

The issue is also present for the `<TermTokenField />` - using it on wpcalypso you can not currently unset all prior-saved tags.

In 6c30400 I implemented a bit of a hacky fix for the problem by sending an array of `[ 0 ]` when all terms have been unset for a given taxonomy.  Additionally I added logic to skip over `post_tag` and `category` taxonomies in the `getTermsById` normalizer logic as it was causing unsavory results in category/tag in the post editor.

aduth commented:
Ref: 9667-gh-calypso-pre-oss

timmyc commented:
😮 at the memory @aduth!  That is funny, the issue felt familiar to me - I looked at the API logic to try to jog my memory but couldn't shake it out.  Moved to using `null` instead and it too works.

aduth commented on client/lib/posts/post-edit-store.js:
```diff
@@ -96,6 +98,22 @@ function getCategoryIds( post ) {
+	}
+
+	// Skip "default" taxonomies
+	const taxonomies = omit( post.terms, [ 'post_tag', 'category' ] );
```
Curious: Does it hurt anything to not omit them here?

timmyc replied:
Ha - yes I should have explained that a bit more.  Currently both `category` and `post_tag` components in the editor sidebar update their perspective attributes on the post object, and do not update the `terms` object.  Since this logic here pulls the term ids from `terms`, and the order of the term logic in the API, if we do not filter out `post_tag` and `category` changes are not saved properly.

Longer-term I can see us moving away entirely from the existing token-field and category-selector implementations at which time we can remove this omit statement and all the data can live in `terms`.

aduth replied:
>Longer-term I can see us moving away entirely from the existing token-field and category-selector implementations at which time we can remove this omit statement and all the data can live in `terms`.

💯  to this

aduth commented on client/post-editor/editor-drawer/taxonomies.jsx:
```diff
@@ -24,35 +25,26 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
+						title={ label }
+						icon={ <Gridicon icon={ icon } /> }
+					>
+					{ hierarchical
```
Now I'm just being obnoxious, but because the components take the same props, you could assign the component to be rendered into a variable:

```diff
diff --git a/client/post-editor/editor-drawer/taxonomies.jsx b/client/post-editor/editor-drawer/taxonomies.jsx
index fd1f0f8..0d1fc04 100644
--- a/client/post-editor/editor-drawer/taxonomies.jsx
+++ b/client/post-editor/editor-drawer/taxonomies.jsx
@@ -32,6 +32,7 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
                                }
 
                                const icon = hierarchical ? 'folder' : 'tag';
+                               const Selector = hierarchical ? TermSelector : TermTokenField;
 
                                return (
                                        <Accordion
@@ -39,10 +40,7 @@ function EditorDrawerTaxonomies( { siteId, postType, postTerms, taxonomies } ) {
                                                title={ label }
                                                icon={ <Gridicon icon={ icon } /> }
                                        >
-                                       { hierarchical
-                                               ? <TermSelector postTerms={ postTerms } taxonomyName={ name } />
-                                               : <TermTokenField postTerms={ postTerms } taxonomyName={ name } />
-                                       }
+                                               <Selector postTerms={ postTerms } taxonomyName={ name } />
                                        </Accordion>
                                );
                        } ).filter( Boolean ) }
```

nylen replied:
This would interfere with #6548 which removes the `postTerms` prop from `TermTokenField`.

aduth commented:
>😮 at the memory @aduth! 

The search-fu is strong in me! 😄

aduth commented:
Seems to be working well. Confirmed that terms are persisted in adding a new term to a new post and removing all terms from existing post. 👍
